### ISSUE TITLE
Using etl/string.h gives basic_string.h:95:26: error: ‘ETL_STRING_FILE_ID’ was not declared in this scope

### ISSUE THREAD

**Issue by bolry** [ETLCPP/etl#348]
Just include <etl/string.h> and you get basic_string.h:95:26: error: ‘ETL_STRING_FILE_ID’ was not declared in this scope in C++98/03 mode.

Using etl_profile.h with content: `#include <etl/profiles/cpp03.h>`

g++-10  -I. -isystem etl-src/include -std=gnu++98 -c string.h.t.cpp

Similar problem with array_wrapper.h ==> error: ‘ETL_ARRAY_WRAPPER_FILE_ID’ was not declared in this scope

**Comment by jwellbelove** [comment]
That's annoying, none of my compilers picked that up and neither did the Github CI!

**Comment by bolry** [comment]
I've made one test with multiple files where each header has it's own cpp-file just to see if it still works in C++98/03 mode. It's been "rewarding". :-)

**Comment by jovere** [comment]
It looks like this is all due to commit 87554af4f8a2b8743d729b911d5723b92f7a031e.  What I don't understand is that if I build the tests, it works fine, but if I just build a simple file with the following contents, it breaks just by including the header:
```c++
#include "etl/string.h"

int main()
{

}
```
Note: I first saw this when trying to use etl::strings in compiler explorer.
<details><summary>Compiler Output (expand)</summary>

```
$ gcc test.cpp -Ietl/include
In file included from etl/include/etl/array.h:45:0,
                 from etl/include/etl/basic_string.h:46,
                 from etl/include/etl/string.h:35,
                 from test.cpp:1:
etl/include/etl/basic_string.h: In constructor ‘etl::string_empty::string_empty(etl::exception::string_type, etl::exception::numeric_type)’:
etl/include/etl/basic_string.h:95:57: error: ‘ETL_STRING_FILE_ID’ was not declared in this scope
       : string_exception(ETL_ERROR_TEXT("string:empty", ETL_STRING_FILE_ID"A"), file_name_, line_number_)
                                                         ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
etl/include/etl/basic_string.h:95:57: note: suggested alternative: ‘ETL_VARIANT_FILE_ID’
       : string_exception(ETL_ERROR_TEXT("string:empty", ETL_STRING_FILE_ID"A"), file_name_, line_number_)
                                                         ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
etl/include/etl/basic_string.h:95:75: error: expected ‘)’ before string constant
       : string_exception(ETL_ERROR_TEXT("string:empty", ETL_STRING_FILE_ID"A"), file_name_, line_number_)
                                                                           ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
In file included from etl/include/etl/string.h:35:0,
                 from test.cpp:1:
etl/include/etl/basic_string.h:97:5: error: no matching function for call to ‘etl::string_exception::string_exception()’
     }
     ^
etl/include/etl/basic_string.h:80:5: note: candidate: etl::string_exception::string_exception(etl::exception::string_type, etl::exception::string_type, etl::exception::numeric_type)
     string_exception(string_type reason_, string_type file_name_, numeric_type line_number_)
     ^~~~~~~~~~~~~~~~
etl/include/etl/basic_string.h:80:5: note:   candidate expects 3 arguments, 0 provided
etl/include/etl/basic_string.h:76:9: note: candidate: constexpr etl::string_exception::string_exception(const etl::string_exception&)
   class string_exception : public etl::exception
         ^~~~~~~~~~~~~~~~
etl/include/etl/basic_string.h:76:9: note:   candidate expects 1 argument, 0 provided
etl/include/etl/basic_string.h:76:9: note: candidate: constexpr etl::string_exception::string_exception(etl::string_exception&&)
etl/include/etl/basic_string.h:76:9: note:   candidate expects 1 argument, 0 provided
etl/include/etl/basic_string.h:97:5: error: expected ‘{’ at end of input
     }
     ^
In file included from etl/include/etl/array.h:45:0,
                 from etl/include/etl/basic_string.h:46,
                 from etl/include/etl/string.h:35,
                 from test.cpp:1:
etl/include/etl/basic_string.h: In constructor ‘etl::string_out_of_bounds::string_out_of_bounds(etl::exception::string_type, etl::exception::numeric_type)’:
etl/include/etl/basic_string.h:109:58: error: ‘ETL_STRING_FILE_ID’ was not declared in this scope
       : string_exception(ETL_ERROR_TEXT("string:bounds", ETL_STRING_FILE_ID"B"), file_name_, line_number_)
                                                          ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
etl/include/etl/basic_string.h:109:58: note: suggested alternative: ‘ETL_VARIANT_FILE_ID’
       : string_exception(ETL_ERROR_TEXT("string:bounds", ETL_STRING_FILE_ID"B"), file_name_, line_number_)
                                                          ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
etl/include/etl/basic_string.h:109:76: error: expected ‘)’ before string constant
       : string_exception(ETL_ERROR_TEXT("string:bounds", ETL_STRING_FILE_ID"B"), file_name_, line_number_)
                                                                            ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
In file included from etl/include/etl/string.h:35:0,
                 from test.cpp:1:
etl/include/etl/basic_string.h:111:5: error: no matching function for call to ‘etl::string_exception::string_exception()’
     }
     ^
etl/include/etl/basic_string.h:80:5: note: candidate: etl::string_exception::string_exception(etl::exception::string_type, etl::exception::string_type, etl::exception::numeric_type)
     string_exception(string_type reason_, string_type file_name_, numeric_type line_number_)
     ^~~~~~~~~~~~~~~~
etl/include/etl/basic_string.h:80:5: note:   candidate expects 3 arguments, 0 provided
etl/include/etl/basic_string.h:76:9: note: candidate: constexpr etl::string_exception::string_exception(const etl::string_exception&)
   class string_exception : public etl::exception
         ^~~~~~~~~~~~~~~~
etl/include/etl/basic_string.h:76:9: note:   candidate expects 1 argument, 0 provided
etl/include/etl/basic_string.h:76:9: note: candidate: constexpr etl::string_exception::string_exception(etl::string_exception&&)
etl/include/etl/basic_string.h:76:9: note:   candidate expects 1 argument, 0 provided
etl/include/etl/basic_string.h:111:5: error: expected ‘{’ at end of input
     }
     ^
In file included from etl/include/etl/array.h:45:0,
                 from etl/include/etl/basic_string.h:46,
                 from etl/include/etl/string.h:35,
                 from test.cpp:1:
etl/include/etl/basic_string.h: In constructor ‘etl::string_iterator::string_iterator(etl::exception::string_type, etl::exception::numeric_type)’:
etl/include/etl/basic_string.h:123:60: error: ‘ETL_STRING_FILE_ID’ was not declared in this scope
       : string_exception(ETL_ERROR_TEXT("string:iterator", ETL_STRING_FILE_ID"C"), file_name_, line_number_)
                                                            ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
etl/include/etl/basic_string.h:123:60: note: suggested alternative: ‘ETL_VARIANT_FILE_ID’
       : string_exception(ETL_ERROR_TEXT("string:iterator", ETL_STRING_FILE_ID"C"), file_name_, line_number_)
                                                            ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
etl/include/etl/basic_string.h:123:78: error: expected ‘)’ before string constant
       : string_exception(ETL_ERROR_TEXT("string:iterator", ETL_STRING_FILE_ID"C"), file_name_, line_number_)
                                                                              ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
In file included from etl/include/etl/string.h:35:0,
                 from test.cpp:1:
etl/include/etl/basic_string.h:125:5: error: no matching function for call to ‘etl::string_exception::string_exception()’
     }
     ^
etl/include/etl/basic_string.h:80:5: note: candidate: etl::string_exception::string_exception(etl::exception::string_type, etl::exception::string_type, etl::exception::numeric_type)
     string_exception(string_type reason_, string_type file_name_, numeric_type line_number_)
     ^~~~~~~~~~~~~~~~
etl/include/etl/basic_string.h:80:5: note:   candidate expects 3 arguments, 0 provided
etl/include/etl/basic_string.h:76:9: note: candidate: constexpr etl::string_exception::string_exception(const etl::string_exception&)
   class string_exception : public etl::exception
         ^~~~~~~~~~~~~~~~
etl/include/etl/basic_string.h:76:9: note:   candidate expects 1 argument, 0 provided
etl/include/etl/basic_string.h:76:9: note: candidate: constexpr etl::string_exception::string_exception(etl::string_exception&&)
etl/include/etl/basic_string.h:76:9: note:   candidate expects 1 argument, 0 provided
etl/include/etl/basic_string.h:125:5: error: expected ‘{’ at end of input
     }
     ^
In file included from etl/include/etl/array.h:45:0,
                 from etl/include/etl/basic_string.h:46,
                 from etl/include/etl/string.h:35,
                 from test.cpp:1:
etl/include/etl/basic_string.h: In constructor ‘etl::string_truncation::string_truncation(etl::exception::string_type, etl::exception::numeric_type)’:
etl/include/etl/basic_string.h:137:60: error: ‘ETL_STRING_FILE_ID’ was not declared in this scope
       : string_exception(ETL_ERROR_TEXT("string:iterator", ETL_STRING_FILE_ID"D"), file_name_, line_number_)
                                                            ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
etl/include/etl/basic_string.h:137:60: note: suggested alternative: ‘ETL_VARIANT_FILE_ID’
       : string_exception(ETL_ERROR_TEXT("string:iterator", ETL_STRING_FILE_ID"D"), file_name_, line_number_)
                                                            ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
etl/include/etl/basic_string.h:137:78: error: expected ‘)’ before string constant
       : string_exception(ETL_ERROR_TEXT("string:iterator", ETL_STRING_FILE_ID"D"), file_name_, line_number_)
                                                                              ^
etl/include/etl/error_handler.h:305:53: note: in definition of macro ‘ETL_ERROR_TEXT’
   #define ETL_ERROR_TEXT(verbose_text, terse_text) (terse_text)   // Use the terse text.
                                                     ^~~~~~~~~~
In file included from etl/include/etl/string.h:35:0,
                 from test.cpp:1:
etl/include/etl/basic_string.h:139:5: error: no matching function for call to ‘etl::string_exception::string_exception()’
     }
     ^
etl/include/etl/basic_string.h:80:5: note: candidate: etl::string_exception::string_exception(etl::exception::string_type, etl::exception::string_type, etl::exception::numeric_type)
     string_exception(string_type reason_, string_type file_name_, numeric_type line_number_)
     ^~~~~~~~~~~~~~~~
etl/include/etl/basic_string.h:80:5: note:   candidate expects 3 arguments, 0 provided
etl/include/etl/basic_string.h:76:9: note: candidate: constexpr etl::string_exception::string_exception(const etl::string_exception&)
   class string_exception : public etl::exception
         ^~~~~~~~~~~~~~~~
etl/include/etl/basic_string.h:76:9: note:   candidate expects 1 argument, 0 provided
etl/include/etl/basic_string.h:76:9: note: candidate: constexpr etl::string_exception::string_exception(etl::string_exception&&)
etl/include/etl/basic_string.h:76:9: note:   candidate expects 1 argument, 0 provided
etl/include/etl/basic_string.h:139:5: error: expected ‘{’ at end of input
     }
     ^
```

</details>

**Comment by jwellbelove** [comment]
I'm working on a fix this weekend + adding bolry's sanity check tests.

**Comment by jwellbelove** [comment]
Fixed in 20.6.2
